README.md
# Socket

![license](https://img.shields.io/badge/license-MIT-brightGreen.svg)
[![build](https://github.com/originphp/socket/workflows/CI/badge.svg)](https://github.com/originphp/socket/actions)
[![coverage](https://coveralls.io/repos/github/originphp/socket/badge.svg?branch=master)](https://coveralls.io/github/originphp/socket?branch=master)

The `Socket` class is a lightweight wrapper which makes working with sockets easy.

## Installation

To install this package

```linux
$ composer require originphp/socket
```

## Using Sockets

```php
use Origin\Socket\Socket;
$socket = new Socket([
    'host' => 'localhost',
    'protocol' => 'tcp',
    'port' => 25,
    'timeout' => 30,
    'persistent' => false,
]);

if ($socket->connect()) {
    $socket->write("HELO mydomain.com\r\n");
    $result = $socket->read();
}

$socket->disconnect();
```

You can can also enable encryption using  `ssl` or `tls`, you can also specify versions e.g. `sslv2`, `sslv23`, `sslv3`, `tlsv1`, `tlsv11` and `tlsv12`. See the [PHP manual](https://www.php.net/manual/en/function.stream-socket-enable-crypto.php) for more information on the encryption methods.

To enable encryption

```php
$socket->enableEncryption('tls');
$socket->enableEncryption('ssl');
```

To disable encryption

```php
$socket->disableEncryption('tls');
$socket->disableEncryption('ssl');
```

## Host/IP Address

To get the IP address of the connection

```php
$ipAddress = $socket->address();
```

To get the hostname

```php
$hostname = $socket->host();
```

## Stream Contexts

When creating a Socket you can also provide `context` options that will be used to create a [stream context](https://www.php.net/manual/en/function.stream-context-create.php).

```php
$socket = new Socket([
    'host' => 'example.com',
    'protocol' => 'tcp',
    'port' => 443,
    'timeout' => 30,
    'persistent' => false,
    'context' => [
        'ssl' => [
            'verify_peer' => false
        ]
    ]
]);
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.0] - 2021-01-04

### Changed

- Changed minimum PHP version to 7.3
- Change minimum PHPUnit to 9.2

## [1.0.1] - 2020-05-25

### Fixed

- Fixed updated tests to work with changes on remote servers

## [1.0.0] - 2019-10-17

This component has been decoupled from the [OriginPHP framework](https://www.originphp.com/).

tests/SocketTest.php
<?php
/**
 * OriginPHP Framework
 * Copyright 2018 - 2021 Jamiel Sharief.
 *
 * Licensed under The MIT License
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * @copyright   Copyright (c) Jamiel Sharief
 * @link        https://www.originphp.com
 * @license     https://opensource.org/licenses/mit-license.php MIT License
 */

namespace Origin\Test\Socket;

use Origin\Socket\Socket;
use InvalidArgumentException;
use Origin\Socket\Exception\SocketException;
use Origin\Socket\Exception\SocketTimeoutException;

class MockSocket extends Socket
{
    public function connection()
    {
        return $this->connection;
    }
}
class SocketTest extends \PHPUnit\Framework\TestCase
{
    public function testConnection()
    {
        $socket = new MockSocket([
            'host' => 'example.com','port' => 80,'timeout' => 5
        ]);
        $this->assertTrue($socket->connect());
        $this->assertIsResource($socket->connection());
        $this->assertTrue($socket->disconnect());
    }

    public function testLastError()
    {
        $socket = new Socket();
        $socket->lastError('foo');
        $this->assertEquals('foo', $socket->lastError());
    }

    public function testConnectionException()
    {
        $this->expectException(SocketException::class);
        $socket = new Socket(['host' => 'foo']);
        $socket->connect();
    }

    public function testConnectionExceptionErrorHandler()
    {
        $this->expectException(SocketException::class);
        $socket = new Socket(['host' => 'www.example.com','protocol' => 'ssl','port' => 80]);
        $socket->connect();
    }

    public function testEnableCryptoException()
    {
        $this->expectException(SocketException::class);
        $socket = new Socket([
            'host' => 'example.com','port' => 80,'timeout' => 5
        ]);
        $this->assertTrue($socket->connect());
        $socket->enableEncryption('tls');
    }

    /**
     * By setting verify_peer to false, i know the ssl context is being set
     *
     * Origin\Socket\Exception\SocketException: stream_socket_client(): SSL operation failed with code 1. OpenSSL Error messages:
    * error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed
    * stream_socket_client(): Failed to enable crypto
     * stream_socket_client(): unable to connect to ssl://originphp.com:443 (Unknown error)
     *
     * @return void
     */
    public function testCreateContext()
    {
        $context = [
            'ssl' => [
                'verify_peer' => false
            ]
        ];
        $socket = new Socket([
            'host' => 'originphp.com','port' => 443,'protocol' => 'ssl','timeout' => 2,
            'context' => $context
        ]);
        $this->assertTrue($socket->connect());
    }

    public function testSocketTimeout()
    {
        $this->expectException(SocketTimeoutException::class);
        $socket = new Socket([
            'host' => 'example.com','port' => 80,'timeout' => 1
        ]);
        $this->assertTrue($socket->connect());
        $socket->read();
    }

    public function testSocketWrite()
    {
        $socket = new Socket([
            'host' => 'example.com','port' => 80,'timeout' => 5
        ]);
        $this->assertEquals(0, $socket->write('It is not connected'));
        $this->assertTrue($socket->connect());
        $this->assertEquals(58, $socket->write("GET http://www.example.com HTTP/1.1\r\nConnection: close\r\n\r\n"));
        $this->assertTrue($socket->disconnect());
    }

    public function testConnectReConnect()
    {
        $socket = new Socket([
            'host' => 'example.com','port' => 80,'timeout' => 5
        ]);
        $this->assertTrue($socket->connect());
        $this->assertTrue($socket->connect());
        $this->assertEquals(58, $socket->write("GET http://www.example.com HTTP/1.1\r\nConnection: close\r\n\r\n"));
        $this->assertTrue($socket->disconnect());
    }

    /**
     * @depends testSocketWrite
     */
    
    public function testSocketRead()
    {
        $socket = new Socket([
            'host' => 'example.com','port' => 80,'timeout' => 5
        ]);
        $this->assertNull($socket->read());
        $this->assertTrue($socket->connect());
            
        $this->assertEquals(58, $socket->write("GET http://www.example.com HTTP/1.1\r\nConnection: close\r\n\r\n"));
       
        $this->assertStringContainsString('<h1>Example Domain</h1>', $socket->read(2000));
        $this->assertNull($socket->lastError());
        $this->assertTrue($socket->disconnect());
    }

    public function testSocketHost()
    {
        $socket = new Socket([
            'host' => 'github.com','port' => 80,'timeout' => 5
        ]);
        $this->assertNull($socket->read());
        $this->assertTrue($socket->connect());
        $this->assertStringContainsString('.github.com', $socket->host());
    }

    public function testSocketAddress()
    {
        $socket = new Socket([
            'host' => 'example.com','port' => 80,'timeout' => 5
        ]);
        $this->assertNull($socket->read());
        $this->assertTrue($socket->connect());
        $this->assertEquals('93.184.216.34', $socket->address());
    }

    public function testSocketSSL()
    {
        $socket = new Socket([
            'host' => 'google.com','port' => 443,'protocol' => 'ssl','timeout' => 5
        ]);
        $this->assertTrue($socket->connect());
        //$this->assertTrue($socket->enableCrypto('sslv3'));
        $this->assertEquals(65, $socket->write("GET https://about.google/intl/en/ HTTP/1.1\r\nConnection: close\r\n\r\n"));
      
        $this->assertStringContainsString('<title>Google - About Google, Our Culture', $socket->read(2500));
    }

    public function testEnableCrypto()
    {
        $socket = new Socket([
            'host' => 'smtp.gmail.com','port' => 587,'timeout' => 5
        ]);
        $this->assertTrue($socket->connect());
        $this->assertStringContainsString('220 ', $this->wait($socket));
        $this->assertTrue((bool) $socket->write("HELO [127.0.0.1]\r\n"));
        $this->assertStringContainsString('250 ', $this->wait($socket));
        
        $this->assertTrue((bool) $socket->write("STARTTLS\r\n"));
        $this->assertStringContainsString('220 ', $this->wait($socket));
        $this->assertTrue($socket->enableEncryption('tls'));
        $this->assertTrue((bool) $socket->write("HELO [127.0.0.1]\r\n"));
        $this->assertStringContainsString('250 ', $this->wait($socket));
        $this->assertTrue($socket->disableEncryption('tls'));
        $this->assertTrue($socket->disconnect());
    }

    public function testEnableCryptoInvalidArgument()
    {
        $this->expectException(InvalidArgumentException::class);
        $socket = new Socket();
        $socket->enableEncryption('ssl-10.0');
    }

    /**
     * This is used by the enableCrypto test
     *
     * @param resource $socket
     * @return string
     */
    private function wait($socket) : string
    {
        $response = '';
        while (true) {
            $buffer = $socket->read();
            if ($buffer === null) {
                break;
            }
            $response .= $buffer;
            if (substr($buffer, 3, 1) == ' ' or strlen($buffer) === 3) {
                break;
            }
        }

        return $response;
    }
}


src/Socket.php
<?php
/**
 * OriginPHP Framework
 * Copyright 2018 - 2020 Jamiel Sharief.
 *
 * Licensed under The MIT License
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * @copyright   Copyright (c) Jamiel Sharief
 * @link        https://www.originphp.com
 * @license     https://opensource.org/licenses/mit-license.php MIT License
 */
declare(strict_types = 1);
namespace Origin\Socket;

use InvalidArgumentException;
use Origin\Socket\Exception\SocketException;
use Origin\Socket\Exception\SocketTimeoutException;

class Socket
{
    /**
     * Holds the configuration
     *
     * @var array
     */
    protected $config = [];

    /**
     * Holds the socket connection
     *
     * @var resource|null
     */
    protected $connection = null;

    /**
     * Holds the last error
     *
     * @var string
     */
    protected $lastError = null;

    /**
     * @var array
     */
    protected $errors = [];

    /**
     * Encryption methods
     *
     * @internal sslv2 not supported . e.g SSLv2 unavailable in this PHP version
     * @see https://www.php.net/manual/en/function.stream-socket-enable-crypto.php
     * @var array
     */
    protected $encryptionMethods = [
        'ssl' => STREAM_CRYPTO_METHOD_SSLv23_CLIENT,
        'sslv2' => STREAM_CRYPTO_METHOD_SSLv2_CLIENT,
        'sslv23' => STREAM_CRYPTO_METHOD_SSLv23_CLIENT,
        'sslv3' => STREAM_CRYPTO_METHOD_SSLv3_CLIENT,
        'tls' => STREAM_CRYPTO_METHOD_TLS_CLIENT,
        'tlsv1' => STREAM_CRYPTO_METHOD_TLSv1_0_CLIENT,
        'tlsv11' => STREAM_CRYPTO_METHOD_TLSv1_1_CLIENT,
        'tlsv12' => STREAM_CRYPTO_METHOD_TLSv1_2_CLIENT,
    ];
    
    public function __construct(array $options = [])
    {
        $options += [
            'host' => 'localhost',
            'protocol' => 'tcp',
            'port' => 80,
            'timeout' => 30,
            'persistent' => false,
            'context' => null // an array
        ];
        $this->config = $options;
    }

    /**
     * Connects to the socket, is a connection is already connected then it will be disconnected
     * first.
     *
     * @return bool
     */
    public function connect() : bool
    {
        if ($this->connection) {
            $this->disconnect();
        }

        set_error_handler([$this, 'errorHandler']);
        $this->connection = stream_socket_client(
            $this->socketAddress(),
            $errorNumber,
            $errorMessage,
            $this->config['timeout'],
            $this->config['persistent']  ? STREAM_CLIENT_PERSISTENT : STREAM_CLIENT_CONNECT, // flags
            stream_context_create($this->config['context'])
        );
        restore_error_handler();

        if (! empty($errorNumber) || ! empty($errorMessage)) {
            $this->lastError("{$errorNumber}:{$errorMessage}");
            throw new SocketException($errorMessage);
        }
    
        /**
         * Throw exception on ErrorHandler errors (can be multiple)
         */
        if (! $this->connection && $this->errors) {
            throw new SocketException(implode("\n", $this->errors));
        }

        $connected = is_resource($this->connection);
        if ($connected) {
            stream_set_timeout($this->connection, $this->config['timeout']);
        }

        return $connected;
    }

    /**
     * Enables Encryption
     *
     * @see https://www.php.net/manual/en/function.stream-socket-enable-crypto.php
     * @param string $type ssl,sslv2,sslv23,sslv3,tls,tlsv1,tlsv11,tlsv12
     * @return boolean
     */
    public function enableEncryption(string $type) : bool
    {
        return $this->enableCrypto($type, true);
    }

    /**
     * Disables Encryption
     *
     * @see https://www.php.net/manual/en/function.stream-socket-enable-crypto.php
     * @param string $type ssl,sslv2,sslv23,sslv3,tls,tlsv1,tlsv11,tlsv12
     * @return boolean
     */
    public function disableEncryption(string $type) : bool
    {
        return $this->enableCrypto($type, false);
    }

    /**
     * Enables (or disables crypto) on a socket
     *
     * @see https://www.php.net/manual/en/function.stream-socket-enable-crypto.php
     * @param string $encryptionMethod ssl,sslv2,sslv23,sslv3,tls,tlsv1,tlsv11,tlsv12
     * @param boolean $enable
     * @return boolean
     */
    private function enableCrypto(string $encryptionMethod, bool $enable = true) : bool
    {
        if (! array_key_exists($encryptionMethod, $this->encryptionMethods)) {
            throw new InvalidArgumentException('Invalid Encryption method.');
        }

        try {
            $result = stream_socket_enable_crypto(
                $this->connection,
                $enable,
                $this->encryptionMethods[$encryptionMethod]
            );
        } catch (\Exception $exception) {
            throw new SocketException($exception->getMessage());
        }

        // returns true, false or 0
        return $result !== 0;
    }

    /**
     * Reads data from the socket
     *
     * @param integer $bytes
     * @return string|null
     * @throws \Origin\Socket\Exception\SocketTimeoutException
     */
    public function read(int $bytes = 1024) : ?string
    {
        if (! $this->connection) {
            return null;
        }
        $buffer = null;
        if (! feof($this->connection)) {
            $buffer = fread($this->connection, $bytes);
            $info = stream_get_meta_data($this->connection);
            if ($info['timed_out']) {
                throw new SocketTimeoutException('Connection timed out.');
            }
        }

        return $buffer;
    }

    /**
    * Writes data to the socket
    *
    * @param string $data
    * @return int
    */
    public function write(string $data) :int
    {
        if (! $this->connection) {
            return 0;
        }

        $bytes = fwrite($this->connection, $data);

        return ($bytes === false) ? 0 : $bytes;
    }

    /**
     * Gets the hostname of the connection
     *
     * @return string
     */
    public function host() : string
    {
        $ipAddress = $this->config['host'];
        if (filter_var($ipAddress, FILTER_VALIDATE_IP) === false) {
            $ipAddress = $this->address();
        }

        return gethostbyaddr($ipAddress);
    }

    /**
     * Gets the IP address of the connection
     *
     * @return string
     */
    public function address() : string
    {
        $ipAddress = $this->config['host'];
        if (filter_var($ipAddress, FILTER_VALIDATE_IP) === false) {
            $ipAddress = gethostbyname($ipAddress);
        }

        return $ipAddress;
    }

    /**
     * Disconnects the current connection
     *
     * @return boolean
     */
    public function disconnect() : bool
    {
        if (is_resource($this->connection) && fclose($this->connection)) {
            $this->connection = null;
        }

        return ($this->connection === null);
    }

    /**
     * Last error setter and getter
     *
     * @param string $error
     * @return string|null
     */
    public function lastError(string $error = null) : ?string
    {
        if ($error === null) {
            return $this->lastError;
        }

        $this->lastError = $error;

        return null;
    }

    /**
     * The error handler used when creating the connection
     *
     * @param int $code
     * @param string $message
     * @return void
     */
    public function errorHandler(int $code, string $message) : void
    {
        $this->errors[] = $message;
    }

    /**
     * Builds the socket address
     *
     * @return string
     */
    private function socketAddress() : string
    {
        $scheme = $this->config['protocol'] ?  $this->config['protocol'] . '://' : null;

        return $scheme . $this->config['host'] . ':' . $this->config['port'];
    }
}
